### DISCUSSION TITLE
Do not specify precision for postgres timestamp fields

### DISCUSSION THREAD

**Discussion by FelipeEmerim** [mikro-orm/mikro-orm#1694] [Ideas]
Hi,
I do not know if this is the right place for this issue, as it may be a driver issue when converting types. Or maybe the javascript Date type cannot hold a date as detailed as postgres so the following problem does not make a difference.

**Describe the bug**
The [postgres wiki](https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_timestamp.280.29_or_timestamptz.280.29) explicitly states that we should never use that because of time inconsistencies. Instead, we should not specify a precision and let the database handle it, only truncating the field when we need.

**To Reproduce**
1. Create an entity
```typescript
import {
  EntityRepositoryType,
  PrimaryKey,
  Entity,
  Property,
} from '@mikro-orm/core';
// eslint-disable-next-line import/no-cycle
import { UserRepository } from '../repositories/user.repository';

@Entity()
export class User {
  [EntityRepositoryType]?: UserRepository;

  @PrimaryKey()
  id: number;

  @Property({ name: 'first_name', type: 'text' })
  firstName: string;

  @Property({ name: 'last_name', type: 'text' })
  lastName: string;

  @Property({ name: 'password', type: 'text' })
  password: string;

  @Property({ name: 'is_active', default: true })
  isActive: boolean;

  @Property({
    name: 'created_at',
    onCreate: () => new Date(new Date().toUTCString()),
  })
  createdAt?: Date;

  @Property({
    name: 'updated_at',
    onUpdate: () => new Date(new Date().toUTCString()),
  })
  updatedAt: Date;
}
```
2. Generate a migration using the
3. The generated migration will be as follows:
```typescript
import { Migration } from '@mikro-orm/migrations';

export class Migration20210416134216 extends Migration {
  async up(): Promise<void> {
    this.addSql(`
      create table "user" (
        "id" serial primary key,
        "first_name" text not null,
        "last_name" text not null,
        "password" text not null,
        "is_active" bool not null default true,
        "created_at" timestamptz(0) not null,
        "updated_at" timestamptz(0) not null
      );,
    `);
  }
}

```

**Expected behavior**
The migration to be generated like this:
```typescript
import { Migration } from '@mikro-orm/migrations';

export class Migration20210416134216 extends Migration {
  async up(): Promise<void> {
    this.addSql(`
      create table "user" (
        "id" serial primary key,
        "first_name" text not null,
        "last_name" text not null,
        "password" text not null,
        "is_active" bool not null default true,
        "created_at" timestamptz not null,
        "updated_at" timestamptz not null
      );,
    `);
  }
}

```

**Versions**

```json
"dependencies": {
    "@elastic/elasticsearch": "^7.12.0",
    "@mikro-orm/cli": "^4.5.3",
    "@mikro-orm/core": "^4.5.3",
    "@mikro-orm/migrations": "^4.5.3",
    "@mikro-orm/nestjs": "^4.2.0",
    "@mikro-orm/postgresql": "^4.5.3",
    "@nestjs/common": "^7.6.15",
    "@nestjs/config": "^0.6.3",
    "@nestjs/core": "^7.6.15",
    "@nestjs/platform-express": "^7.6.15",
    "@nestjs/sequelize": "^0.2.0",
    "@nestjs/swagger": "^4.8.0",
    "bcrypt": "^5.0.1",
    "class-transformer": "^0.4.0",
    "class-validator": "^0.13.1",
    "common-tags": "^1.8.0",
    "dotenv": "^8.2.0",
    "elastic-apm-node": "^3.13.0",
    "nest-winston": "^1.4.1",
    "pg": "^8.6.0",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^6.6.7",
    "sequelize": "^6.6.2",
    "sequelize-typescript": "^2.1.0",
    "sinon": "^10.0.0",
    "swagger-ui-express": "^4.1.6",
    "winston": "^3.3.3",
    "winston-elasticsearch": "^0.15.4"
  },
  "devDependencies": {
    "@nestjs/cli": "^7.6.0",
    "@nestjs/schematics": "^7.3.1",
    "@nestjs/testing": "^7.6.15",
    "@types/bcrypt": "^3.0.1",
    "@types/chai": "^4.2.16",
    "@types/common-tags": "^1.8.0",
    "@types/express": "^4.17.8",
    "@types/faker": "^5.5.1",
    "@types/jest": "^26.0.22",
    "@types/node": "^14.14.37",
    "@types/sequelize": "^4.28.9",
    "@types/sinon": "^10.0.0",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^4.22.0",
    "@typescript-eslint/parser": "^4.22.0",
    "chai": "^4.3.4",
    "eslint": "^7.24.0",
    "eslint-config-airbnb-base": "^14.2.1",
    "eslint-config-airbnb-typescript": "^12.3.1",
    "eslint-config-airbnb-typescript-prettier": "^4.2.0",
    "eslint-config-prettier": "^8.2.0",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-prettier": "^3.1.4",
    "faker": "github:marak/faker.js",
    "jest": "^26.6.3",
    "prettier": "^2.1.2",
    "supertest": "^6.0.0",
    "ts-jest": "^26.5.4",
    "ts-loader": "^8.1.0",
    "ts-node": "^9.0.0",
    "tsconfig-paths": "^3.9.0",
    "typescript": "^4.2.4"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": ".",
    "testRegex": ".*\\.spec\\.ts$",
    "testPathIgnorePatterns": [
      ".*\\e2e\\.spec\\.ts$"
    ],
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  },
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts",
      "./dist/mikro-orm.config.js"
    ]
  }
}
```
I know I can use columnType and specify the type I want, but since it is a postgres standard I think you could use the correct type by default.

Thank you for the attention and for making this amazing ORM.

**Comment by B4nan** [comment]
In JS it is not possible to have microseconds (which is the default in pg). If we use the default precision, which in pg is microseconds, the change set tracking would be triggered every time, as the value on runtime would be trimmed to milliseconds. 

In v5 the default length is `3` so it supports milliseconds by default. I was thinking about how to support microseconds somehow, at least for reading, but want to finish with the huge schema diffing refactoring first (#1641). It would require to force returning strings for datetime columns, and we would need to handle the conversion manually. And ideally this should work in mysql too.

I was thinking about extending the native `Date` object, adding something like `_raw` property, that would hold the full string variant obtained from the db - and use that for change set tracking. But still we'd have to disable the automatic conversion of datetime strings to `Date` objects, which is quite handy, and it would also have (breaking) effect on how the driver API works (and QB too, which might be bigger issue).

**Comment by FelipeEmerim** [comment]
Thank you for the quick response!

Based on what you said I believe this to be way too much work for such a small gain. With the support milliseconds you are probably  attending the vast majority of scenarios since Javascript itself does not support microseconds. 

I also read the docs (which are really good) and found that it is possible to use @Formula to get the date as string on another field if needed. There are also custom types which could work if on converting to rawSQL we pass a function that converts a date string to timestamptz and on retrieving a function that parses the date into a string.
